File: src/leetcode/python/greedy/WiggleSubsequence_376.py
class WiggleSubsequence_376(object):
    def wiggleMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        if length == 0:
            return 0

        up, down = False, False
        previou = nums[0]
        max_len = 1
        i = 1
        while i < length:
            if nums[i] - previou > 0 and not up:
                while i < length and nums[i] > nums[i - 1]:
                    i += 1
                i -= 1
                previou = nums[i]
                up = True
                max_len += 1
                if up and down:
                    down = False
            elif nums[i] - previou < 0 and not down:
                while i < length and nums[i] < nums[i - 1]:
                    i += 1
                i -= 1
                previou = nums[i]
                down = True
                max_len += 1
                if up and down:
                    up = False
            i += 1
        return max_len

File: src/leetcode/java/greedy/WiggleSubsequence_376.java
package leetcode.java.greedy;

public class WiggleSubsequence_376 {
  public int wiggleMaxLength(int[] nums) {
    if (nums.length == 0) return 0;

    int length = 1;
    boolean up = false, down = false;
    int previous = nums[0];
    for (int i = 1; i < nums.length; i++) {
      if (nums[i] - previous > 0 && !up) {
        while (i < nums.length && nums[i] > nums[i - 1]) {
          i++;
        }
        i--;
        previous = nums[i];
        up = true;
        length++;
        if (up && down) down = false;
      } else if (nums[i] - previous < 0 && !down) {
        while (i < nums.length && nums[i] < nums[i - 1]) {
          i++;
        }
        i--;
        previous = nums[i];
        down = true;
        length++;
        if (up && down) up = false;
      }
    }
    return length;
  }
}
